Fixes #2727 - Rename issues invalid name clash error

BZngr opened the pull request:
This PR fixes #2727.  Includes (previously) failing unit test replicating the #2727 issue.  Adds a pubic static class 'AccessibilityEvaluator' (better name suggestions welcomed) based on the accessibility Scope function from PR #2784 (AssignedByValParameterMakeLocalCopyQuickFix).  If this approach to correcting #2727 is accepted, I would update AssignedByValParameterMakeLocalCopyQuickFix to use this same static class).  The prior Scope resolution function in RefactorRename.cs (FindDeclarationForIdentifier()) contained the bug causing issue #2727... "item.Scope.Contains(_model.Target.Scope)"...in the first LINQ expression.  That is, "TestModule1.IndustryCode" 'Contains' "TestModule1.Industry" scopes which satisfied the query condition and introduced a false-positive collision.  I also tested the FindDeclarationForIdentifier() code against the accessibility resolution tests in AssignedByValParameterMakeLocalCopyQuickFixTests and it resulted in test failures there as well.  So, that is why I elected to replace FindDeclarationForIdentifier() in total.  It seems possible that the static class 'AccessibilityEvaluator.GetDeclarationsAccessibleToScope( Declaration target, IEnumerable<Declaration> declarations)' function could optionally be exposed by RubberduckParserState. Exposing the function there may reduce fetches of the AllUserDeclarations list.  AllUserDeclarations would typically be the List passed into the new function - but I kept ParserState out of the function parameters so any IEnumermable<Declaration> (however derived) could be used.

retailcoder commented:
I think the scope evaluation code belongs in the DeclarationFinder more than in a static class; that way anything that can access parser state can access the functionality, without introducing implicit dependencies like a static class does.

BZngr commented:
OK...I'm still learning the solution's class organization etc, so I'm happy to put it in DeclarationFinder.  I just wanted to make sure I could re-use it.

comintern commented:
Just remember that anything in the `DeclarationFinder` has to be thread safe.

BZngr commented:
Moved code to DeclarationFinder.  It's been a long time since I've wandered outside the STA, and never before in C# - so please take a look in DeclarationFinder.  Since the code I introduced does not modify any instance variables and everything is on the stack, I'm wondering if the 'lock' is actually needed.

retailcoder commented:
Nah, lock would be for writing to instance state - should be good without.

BZngr commented:
Lock removed.

retailcoder commented on RubberduckTests/Refactoring/RenameTests.cs:
```diff
@@ -21,6 +21,74 @@ namespace RubberduckTests.Refactoring
+            var qualifiedSelection = new QualifiedSelection(new QualifiedModuleName(component), selection);
+
+            var msgbox = new Mock<IMessageBox>();
+            msgbox.Setup(m => m.Show(It.IsAny<string>(), It.IsAny<string>(), MessageBoxButtons.YesNo, It.IsAny<MessageBoxIcon>()))
```
You've set up the mock - if the test means to verify that msgbox.show was never invoked, the `msgbox.Verify()` condition failing would be enough to fail the test- no need to `Assert` if you `Verify` ;-)

BZngr replied:
I used the Assert.IsFalse so I could provide a friendlier message.  And, more importantly, the exception on msgbox.Verify() is the success path - so I think I have to catch it.

retailcoder replied:
Add another dot adter the verify call; you can specify how many times you're expecting the call to be made, here 0.
